pkg/deployment/cert.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployment
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	certmgrv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	certmgrmetav1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	dataplanev1 "github.com/openstack-k8s-operators/dataplane-operator/api/v1beta1"
+	infranetworkv1 "github.com/openstack-k8s-operators/infra-operator/apis/network/v1beta1"
+	"github.com/openstack-k8s-operators/lib-common/modules/certmanager"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/secret"
+)
+
+// GenerateTLSCerts generates secrets containing certificates for relevant services
+func GenerateTLSCerts(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1.OpenStackDataPlaneNodeSet,
+	allIPSets map[string]infranetworkv1.IPSet) error {
+
+	// We need to create a secret for each service that contains all the certs needed for
+	// that service for all the nodes in the nodeset.  This one secret will be mounted in
+	// the ansibleEE pod as an extra mount when that service is deployed.
+	//
+	// The convention we use here is "<nodeset.name>-<service>-certs", so for example,
+	// openstack-epdm-nova-certs.
+	novaCertsData := map[string][]byte{}
+	libvirtCertsData := map[string][]byte{}
+
+	// for each node in the nodeset, issue all the TLS certs needed based on the
+	// ips or DNS Names
+	for nodeName, _ := range instance.Spec.Nodes {
+		var ips []string
+		var dnsNames []string
+		var secretName string
+		var certName string
+
+		// TODO(alee) decide if we want to use other labels
+		// For now we just add the hostname so we can select all the certs on one node
+		labels := map[string]string{
+			"hostname": nodeName,
+		}
+
+		ipSet, ok := allIPSets[nodeName]
+		if ok {
+			for _, res := range ipSet.Status.Reservation {
+				ips = append(ips, res.Address)
+				fqdnName := strings.Join([]string{nodeName, res.DNSDomain}, ".")
```

Comment:
These are already available generated for dnsdata[1], we probbaly should not go through all IPSets and build them again. Can return a dict from there that can be used in nodeset controller when calling GenerateTLSCerts()

[1] https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/pkg/deployment/ipam.go#L87

Reply:
Yup - this is where I originally saw the code I used.

I'd have to return all the dnsNames and all the ips for each node - because its likely we'll need both.
We can do as you suggest, but it does make the code a little less readable IMHO.  Is it really that much of a performance improvement to go through the ips again?

Reply:
It's kind of duplicating the code. Also, using the `nodeName` to get the fqdn is not correct and we've changed that[1].  As a minimum we should have a helper function that's used at both places so that we keep them consistent.

[1] https://github.com/openstack-k8s-operators/dataplane-operator/commit/14f9f9b236276d00d53d3991b7537faf9498fe44

Reply:
Ok - I see that this code can diverge -- I've taken your suggestion in the next iteration.